### PR TITLE
feat(ST-65): implement reset password endpoint

### PR DESCRIPTION
**Pull request by ghost**
## Description of Change
- Created Otp module, service and repository for generating OTP and validating OTP
- Updated auth controller and service for implementing 2 endpoint, one is for generating OTP and one for reseting password

## Associated Ticket Link(s)
- https://trello.com/c/o7XmJKp3

## Screenshots / Screen Recordings
N/A

### CODE REVIEW

File: src/auth/auth.controller.ts [hunk 1]
```diff
@@ -23,4 +23,18 @@ export class AuthController {
...
+  @Post("reset-password/generate-otp")
+  generateResetPasswordOtp(@Body("email") email: string): Promise<string> {
+    return this.authService.generateResetPasswordOtp(email);
+  }
+
+  @Post("reset-password")
+  resetPassword(
+    @Body("email") email: string,
+    @Body("otp") otpCode: string,
+    @Body("newPassword") newPassword: string,
```
Comment (ishmam-iqbal-sazim) [inline src/auth/auth.controller.ts]: Does it make sense to use a dto class and pass the dto itself for the whole Body object?


```suggestion
    @Body() resetPasswordDto: ResetPasswordDto
```

File: src/otp/otp.service.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  private generateRandomOtp(): string {
+    return crypto.randomBytes(3).toString("hex").toUpperCase();
+  }
+
+  async generateOtp(email: string): Promise<string> {
+    const existingOtp = await this.otpRepository.findOne({ email });
+
+    if (existingOtp) {
+      await this.otpRepository.removeOne(existingOtp);
+    }
```
Comment (ishmam-iqbal-sazim) [inline src/otp/otp.service.ts]: Is it necessary to remove it, wouldn't we have to do one less trip to the database if we just updated existing records?

File: src/auth/auth.service.ts [hunk 3]
```diff
@@ -92,4 +96,29 @@ export class AuthService {
...
+
+    return otpCode;
+  }
+
+  async resetPassword(email: string, otpCode: string, newPassword: string): Promise<boolean> {
+    await this.otpService.validateOtp(email, otpCode);
+
+    const user = await this.userService.findUserByEmail(email);
+    const hashedPassword = await bcrypt.hash(newPassword, 10);
+    user!.password = hashedPassword;
```
Comment (ishmam-iqbal-sazim) [inline src/auth/auth.service.ts]: I think it's best to use assign here.
Also, let's avoid non null assertions `!` unless absolutely necessary

File: src/otp/otp.service.ts (hunk 2 above)
Comment (ghost) [reply src/otp/otp.service.ts]: Thank you for the suggestion, I will do that.

**Review by ragib-hossain-sazim — APPROVED** [review]

**Pull request merged**